portfolio/src/test/java/com/google/sps/NumCommentsTest.java

```diff
@@ -0,0 +1,200 @@
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
+import com.google.sps.data.Comment;
+import javax.servlet.http.*;
+import java.io.*;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@RunWith(JUnit4.class)
+public class NumCommentsTest {
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+  
+  /*
+   * Helper that simply removes all whitespace and invisible chars (tabs, \n...)
+   */
+  public String removeWhitespace(String input) {
+    return input.replaceAll("\\s+","");
+  }
+
+  /*
+   * Turn a response StringWriter into an Int
+   */
+  public int StringWriterToInt(StringWriter stringWriter) {
+    String string = stringWriter.toString();
+    String withoutWhitespace = removeWhitespace(string);
+    try {
+      int inputValue = Integer.valueOf(removeWhitespace(string));
+      return inputValue;
+    } catch (NumberFormatException nfe) {
+      throw new NumberFormatException("Couldn't parse " + withoutWhitespace);
+    }
+  } 
+
+  /*
+   * Just make a basic comment entity for testing purposes.
+   */
+  public Entity basicComment() {
+    Entity commentEntity = new Entity("Comment");
+    Long oneLong = Long.valueOf(1);
+
+    commentEntity.setProperty("comment", "");
+    commentEntity.setProperty("timestamp", oneLong);
+    commentEntity.setProperty("upvoteCount", oneLong);
+    commentEntity.setProperty("id", oneLong);
+    return commentEntity;
+  }
+
+  /*
+   * 
+   */
+  public StringWriter getStringWriter(HttpServletResponse response) throws Exception {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+    
+    return stringWriter;
+  }
+
+  @Test
+  public void testEmpty() throws Exception {
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    
+    HttpServletRequest request = mock(HttpServletRequest.class);       
+    HttpServletResponse response = mock(HttpServletResponse.class); 
+
+    StringWriter stringWriter = getStringWriter(response);
+
+    new NumCommentsServlet().doGet(request, response);
+    
+    int numComments = StringWriterToInt(stringWriter);
+    Assert.assertEquals(numComments, 0);
+  }
+
+  @Test 
+  public void testOneComment() throws Exception {
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
```

Comment:
Should definitely be mocking this.

In your header:

@Mock
DatastoreService mockDatastoreService;

In your test:

when(mockDatastoreService).getComments(query).thenReturn(listOfComments)
or whatever the method signature is.